Sources/DarkModeCore/DMTraitCollection.m

```diff
@@ -4,17 +4,127 @@
+
+@import ObjectiveC;
+
+@implementation NSObject (DMTraitEnvironment)
+
++ (void)swizzleTraitCollectionDidChangeToDMTraitCollectionDidChange {
+  [self swizzleTraitCollectionDidChangeToDMTraitCollectionDidChangeWithBlock:nil];
+}
+
++ (void)swizzleTraitCollectionDidChangeToDMTraitCollectionDidChangeWithBlock:(void (^)(NSObject *, UITraitCollection *))block API_AVAILABLE(ios(13.0)) {
+  // Only swizzling classes that conforms to both UITraitEnvironment & DMTraitEnvironment
+  if (!class_conformsToProtocol(self, @protocol(UITraitEnvironment)) || !class_conformsToProtocol(self, @protocol(DMTraitEnvironment))) {
+    return;
+  }
+
+  SEL selector = @selector(traitCollectionDidChange:);
+  Method method = class_getInstanceMethod(self, selector);
+
+  if (!method)
+    NSAssert(NO, @"Method not found for [%@ traitCollectionDidChange:]", NSStringFromClass(self));
+
+  IMP imp = method_getImplementation(method);
+  class_replaceMethod(self, selector, imp_implementationWithBlock(^(NSObject *self, UITraitCollection *previousTraitCollection) {
+    // Call previous implementation
+    ((void (*)(NSObject *, SEL, UITraitCollection *))imp)(self, selector, previousTraitCollection);
+
+    // Call DMTraitEnvironment
+    [(id <DMTraitEnvironment>)self dmTraitCollectionDidChange:previousTraitCollection == nil ? nil : [DMTraitCollection traitCollectionWithUITraitCollection:previousTraitCollection]];
+
+    // Call custom block
+    if (block) {
+      block(self, previousTraitCollection);
+    }
+  }), method_getTypeEncoding(method));
+}
+
+@end
 
 @implementation DMTraitCollection
 
-static DMTraitCollection *_currentTraitCollection = nil;
+static DMTraitCollection *_lastManuallySetTraitCollection = nil; // This is set manually in setCurrentTraitCollection:animated
+static __weak UIApplication *_registeredApplication = nil;
 
 + (DMTraitCollection *)currentTraitCollection {
-  return _currentTraitCollection;
+  DMTraitCollection *lastTraitCollection = [self lastManuallySetTraitCollection];
+  if (@available(iOS 13.0, *)) {
+    if (lastTraitCollection.userInterfaceStyle == DMUserInterfaceStyleUnspecified) {
+      // Return the ones used from the system when no specific user interface style is explicitly set
+      return [self currentSystemTraitCollection];
+    }
+  }
+  return lastTraitCollection;
+}
+
++ (DMTraitCollection *)lastManuallySetTraitCollection {
+  if (!_lastManuallySetTraitCollection) {
+    // Provide unspecified at first
+    _lastManuallySetTraitCollection = [DMTraitCollection traitCollectionWithUserInterfaceStyle:DMUserInterfaceStyleUnspecified];
+  }
+  return _lastManuallySetTraitCollection;
 }
 
-+ (void)setCurrentTraitCollection:(DMTraitCollection *)currentTraitCollection {
-  _currentTraitCollection = currentTraitCollection;
++ (DMTraitCollection *)currentSystemTraitCollection API_AVAILABLE(ios(13.0)) {
+  return [DMTraitCollection traitCollectionWithUITraitCollection:UIScreen.mainScreen.traitCollection];
```

Comment:
I think this will break view/view controller's `overrideUserInterfaceStyle`.

Reply:
no it will not, tested that view/view controller's overrideUserInterfaceStyle is not be affected by its superview/parent view controller.